### DISCUSSION TITLE
Change object-or-collection-initializer to object-and-collection-initializer

### DISCUSSION THREAD

**Discussion by mariusGundersen** [dotnet/csharplang#597] [General]
According to [this SO answer](http://stackoverflow.com/a/6254225/1585) it's not possible to combine object initializer with collection initializer. I have a situation where this would be useful.

First, lets have a look at just the collection initializer. It uses some simple pattern matching to find an `Add()` method with the correct signature, and is even able to work with extension methods. This lets me create this very useful extension method:

```
public static void Add<T>(this ICollection<T> destination, IEnumerable<T> source)
{
    foreach (var item in source)
    {
        destination.Add(item);
    }
}
```

Using this it's now possible to create a new `ICollection` and copy the contents of another `IEnumerable`:

```
new List<string>
{
    { "Prefix" },
    someOtherList,
    { "Suffix" }
};
```

With a dictionary I would like to use the index initializer:
```
var dictionary = new Dictionary<string, string>
{
    someOtherDictionary,
    ["Alpha"] = "one",
    ["Beta"] = "two"
};
```

Unfortunately this is not possible, since the `["Alpha"] = "one"` is using the object initializer. It is possible to do this using the existing collection initializer:

```
var dictionary = new Dictionary<string, string>
{
    someOtherDictionary,
    { "Alpha", "one"},
    { "Beta", "two"}
}
```

But this is not as nice.

**Comment by svick** [comment]
For `Dictionary<K, V>`, you can already write the following code instead:

```c#
var dictionary = new Dictionary<string, string>(someOtherDictionary)
{
    ["Alpha"] = "one",
    ["Beta"] = "two"
};
```

Is this also "not as nice"?

**Comment by mariusGundersen** [comment]
That's my point, you can't combine that with a property, which is what I'm asking for.
